src/java.base/share/classes/java/lang/template/TemplateRuntime.java

```diff
@@ -0,0 +1,392 @@
+        }
+        for (Object value : st.values()) {
+            result.add(value == null ? Object.class : value.getClass());
+        }
+        return result;
+    }
+
+    /**
+     * Return {@link MethodHandle MethodHandles} to access a
+     * {@link StringTemplate StringTemplate's} values.
+     *
+     * @param st  StringTemplate to examine
+     *
+     * @return list of {@link MethodHandle MethodHandles}
+     *
+     * @throws NullPointerException if st is null
+     *
+     * @implNote The default method determines if the {@link StringTemplate}
+     * was synthesized by the compiler, then the MethodHandles are precisely those of the
+     * embedded expressions fields, otherwise this method returns getters for the values list.
+     */
+    static List<MethodHandle> valueGetters(StringTemplate st) {
+        Objects.requireNonNull(st, "st must not be null");
+        List<MethodHandle> result = new ArrayList<>();
+        Class<?> tsClass = st.getClass();
+        if (tsClass.isSynthetic()) {
+            try {
+                for (int i = 0; ; i++) {
+                    Field field = tsClass.getDeclaredField("x" + i);
+                    result.add(JLIA.unreflectField(field, false));
+                }
+            } catch (NoSuchFieldException ex) {
+                // End of fields
+            } catch (ReflectiveOperationException | SecurityException ex) {
+                throw new InternalError(ex);
+            }
+
+            return result;
+        }
+        try {
+            MethodHandles.Lookup lookup = MethodHandles.lookup();
+            MethodHandle getter = lookup.findStatic(TemplateRuntime.class, "getValue",
+                MethodType.methodType(Object.class, int.class, StringTemplate.class));
+            int size = st.values().size();
+            for (int index = 0; index < size; index++) {
+                result.add(MethodHandles.insertArguments(getter, 0, index));
+            }
+            return result;
+        } catch (ReflectiveOperationException | SecurityException ex) {
+            throw new InternalError(ex);
+        }
+    }
+
+    /**
+     * Private ethod used by {@link TemplateRuntime#valueGetters(StringTemplate)}
+     * to access values.
+     *
+     * @param index values index
+     * @param st    the {@link StringTemplate}
+     *
+     * @return value at index
+     */
+    private static Object getValue(int index, StringTemplate st) {
+        return st.values().get(index);
+    }
+
+    /**
+     * Creates a string that interleaves the elements of values between the
+     * elements of fragments.
+     *
+     * @param fragments  list of String fragments
+     * @param values     list of expression values
+     *
+     * @return String interpolation of fragments and values
+     *
+     * @throws NullPointerException fragments or values is null or if any of the fragments is null
+     */
+    static String interpolate(List<String> fragments, List<Object> values) {
```

Comment:
I think it should be better to ensure that the caller always call with a List.of() or a List.copyOf() so null is not a possible element

Reply:
Changing